### PR TITLE
Replace Version.CURRENT toString usages with Build.current() equivalents

### PR DESCRIPTION
**Pull request by ldematte**
Several uses of `Version.CURRENT` just want to get a String representation, for example to send to log or build an Exception message. `Build.current()` has the current version but as a String, and that is replaced in serverless to be the build hash instead of a semantic version. 

This PR replaces some uses of `Version.CURRENT.toString()` (and those that implicitly convert to a String) to use `Build.current().version()`.
It also updates some out-of date comments mentioning Version.CURRENT

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java:116]: @thecoop I rephrased this to avoid an explicit version; do you think it might work or better to revert and just use majors here?

File: server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java:37]: The change here tries to keep the current format close to the one today, checking if `version()` contains a semantic version or just an opaque string (e.g. in serverless). Let me know if it makes sense, or if we should just remove version from the warn agent, or if we should relax the check and allow basically any string for the "version" part of the agent.

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java
Comment (ldematte) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:315]: @williamrandolph I see you have made exactly the same change; I will merge your PR in and adjust once approved.

File: x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlQueryRequest.java
Comment (ldematte) [inline x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlQueryRequest.java:292]: This will be handled separtely (https://elasticco.atlassian.net/browse/ES-6866)

File: server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java:37]: In principle I'm a little worried about the change here. I'm not sure that we have a good way to unit-test the non-semantic code path within the Elasticsearch code base.

I think I'd be happy with it if we could make the `getPatternWithSemanticVersion` and `getPatternWithoutSemanticVersion` methods visible for testing, and adjust our unit tests to cover both cases.

I don't have any objection to removing the semantic version check part of the regex altogether, but we should see if @pgomulka has concerns about that.

File: server/src/test/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifierTests.java
Comment (williamrandolph) [inline server/src/test/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifierTests.java:121]: My comment in `IndexMetadataVerifier` applies here too: we could call `Build.current().major()` and get an `8` for our message.

**Review by williamrandolph — CHANGES_REQUESTED** [review]
It's looking good overall. I have a few comments and suggestions for the trickier cases.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java:116]: I'm a little reluctant to add a semantic concept like a major version in Build.current(). What will serverless return? Or maybe you are saying that 8 will be the "perpetual major" (the final ES release) and we should never increment it again?

File: server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java:37]: Adding unit tests was definitely my plan, sorry for not mentioning it: I wanted to see if you liked this or not, or if it was better to skip (match all) the semantic version part from the regex. Let me know, I will add tests accordingly

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java:116]: On the other hand, for this case (index compatibility) maybe it makes sense to just leave Version.major: maybe for this kind of things we will always have a concept of compatibility through major versions, and it is fine as-is.

File: server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java:37]: I think it looks good. Let's just make sure that at least semantic path is tested in elasticsearch 'statefull' repo. The opaque string we can test in serverless repo (if needed)

I am not sure what is the future of deprecations in serverless. It wasn't decided yet.

Comment (ldematte) [reply server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java:37]: Thanks for the input, I will keep the change and add tests then

**Comment by ldematte** [comment]
@elasticsearchmachine run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java:37]: I've added a couple of tests here, and prepared a PR for serverless to merge once this is updated/brought into the serveless repo: https://github.com/elastic/elasticsearch-serverless/pull/952 

If we don't like it and we want to be able to run all tests in the ES codebase, we can think about alternatives, finding a way to inject `Build` (via internal methods or via  `BuildExtension` - even if this might be elaborate, see `ExtensionLoaderTests`).
Let me know!

Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java:37]: I think the unit tests in serverless look good. I think it'd be good to put a comment in this codebase along the lines of `// Build.current().version() may not be semantic in downstream projects or future releases` so that no one mistakenly sees this code as needing cleanup.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java:116]: I don't expect 8 to be the perpetual major. My thought was actually that I don't know what we will be doing about major versions, but we could at least centralize any logic for them so that we can make our changes in one place.

What we need to convey here is that if a user is upgrading a node to 8.11 but has an index from 6.x or earlier, they need to first upgrade to 7.x (preferably 7.17.x), and _then_ upgrade to 8.x.

In settings that we're managing, we could probably understand something like "a node that supports org.apache.lucene.util.Version.CURRENT.major - 1", but if I recall correctly the Lucene major and Elasticsearch major are not aligned. That would be confusing for a lot of self-managing users.

I don't really know what to do, but I'd be fine to leave Version.major here for now and come back to it in a later PR.

Comment (ldematte) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java:116]: I agree with you, we probably need a better way to express this information or concept. It may be linked to the "era" concept or features that Simon is working on (e.g. 7.x could be mapped to a `can-reindex-6x`). Also, it could be that moving to IndexVersion changes the definition of index compatibility.
I will revert my change and leave Version.major here, making  a note to  revisit it when we have a clearer direction.
Thanks!

File: server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java:37]: Very good point, will add it

File: server/src/test/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifierTests.java
Comment (ldematte) [reply server/src/test/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifierTests.java:121]: As per the other conversation, we agreed to leave the message as-is until we have a better understanding how major will evolve (also in light of CD projects)

**Review by williamrandolph — APPROVED** [review]
Looks good to me!

**Pull request merged**
